onnxruntime/core/optimizer/propagate_cast_ops.cc

```diff
@@ -64,76 +64,73 @@ static std::string GetName(const std::pair<const NodeArg*, std::vector<Node*>>&
   return p.first->Name() + " feeding " + ConcatNames(p.second) + "; ";
 };
 /*
-* The collection fp16_allow_ops, specifies for a given propagate_cast_ops level, a vector of node op_types that
-* the code is allowed to propage Cast operations across. The user may specify a custom list of optypes using level 0.
-* The opcodes are split into multiple levels. Cast propagation is done based on the level. Level 2 op code
-* list includes Level 1 list also.
-*/
-static std::vector<std::unordered_set<std::string>> fp16_allow_ops = {
+ * The collection fp16_allow_ops, specifies for a given propagate_cast_ops level, a vector of node op_types that
+ * the code is allowed to propage Cast operations across. The user may specify a custom list of optypes using level 0.
+ * The opcodes are split into multiple levels. Cast propagation is done based on the level. Level 2 op code
+ * list includes Level 1 list also.
+ */
+using NodeIndices = std::unordered_set<NodeIndex>;
+using FP16AllowOps = PropagateCastOps::FP16AllowOps;
+static const FP16AllowOps default_fp16_allow_ops = {
     /* Level 0 */ {},
     /* Level 1 */ {"Expand", "Transpose", "Relu", "Reshape", "Split", "Tanh", "Squeeze", "Unsqueeze"},
     /* Level 2 */ {"Add", "BiasGelu", "Dropout", "FastGelu", "Gather", "Gelu", "LayerNormalization", "Where"}};
 
 /*
-*  The following two maps specify the opcode to input and opcode to output mappings to list the inputs/outputs to consider while propagating
-*  cast operations. All other inputs/outputs not listed in this table are not relevant for deciding whether an operation
-*  performed in float or float16. If an opcode is not listed in these tables, the code will look at all the inputs and outputs to validate 
-*  transformation.
-*/
-static std::unordered_map<std::string, std::vector<int>> opcode_to_input_map = {
+ *  The following two maps specify the opcode to input and opcode to output mappings to list the inputs/outputs to consider while propagating
+ *  cast operations. All other inputs/outputs not listed in this table are not relevant for deciding whether an operation
+ *  performed in float or float16. If an opcode is not listed in these tables, the code will look at all the inputs and outputs to validate
+ *  transformation.
+ */
+static const std::unordered_map<std::string, std::vector<int>> opcode_to_input_map = {
```

Comment:
Generally speaking, we should avoid using global  variables who have dynamic initializers. Like std::vector/set/... , the things that can't be initialized at compile time.  For example, here maybe most applications do not need this optimizers, but they still need to run this line of code to init the map. The cost seems small, but if when we put it under the situation that billions of Windows devices are running onnxruntime, it becomes important.

Reply:
This data is not initialized at the load time or at run time. This data goes into the initialized data segment. The compiler will handle it. So, this does not pose any performance penalty.

Reply:
I don't think this is initialized at compile time. unordered_map doesn't have a constexpr constructor.
One alternative is to put it into a function:
```c++
const std::unordered_map<std::string, std::vector<int>>& GetOpCodeToInputMap() {
  static const std::unordered_map<std::string, std::vector<int>> opcode_to_input_map = {...};
  return opcode_to_input_map;
}
```
Or make the data structure constexpr which might be more work.